### PR TITLE
[WIP] Initial implementation for tab stops in TerminalDispatch

### PR DESCRIPTION
**Pull request by skyline75489**
## Summary of the Pull Request

Tired of reviewing those PRs that might cause regression somewhere? Afraid that something bad happens and we might need another servicing release? Seeking for a PR that's totally not useful at this moment? Well, today is your lucky day! Ladies & gentlemen, behold, for a PR that literally only ~10 persons on this planet know & care what it does. 

 
## PR Checklist
* [ ] Closes nothing
* [X] CLA signed. If not, go over [here](https://cla.opensource.microsoft.com/microsoft/Terminal) and sign the CLA
* [ ] Tests added/passed
* [ ] Documentation updated. If checked, please file a pull request on [our docs repo](https://github.com/MicrosoftDocs/terminal) and link it here: #xxx
* [ ] Schema updated.
* [ ] I've discussed this with core contributors already. If not checked, I'm ready to accept this work might be rejected in favor of a different grand plan. Issue number where discussion took place: #xxx

## Detailed Description of the Pull Request / Additional comments

Basically I try to copy the tab handling code inside `adaptDispatch` into `TerminalDispatch` and make it work.

### CODE REVIEW

File: src/cascadia/TerminalCore/TerminalDispatch.cpp
Comment (skyline75489) [inline src/cascadia/TerminalCore/TerminalDispatch.cpp:148]: I need a size which is provided by `_pConApi->GetConsoleScreenBufferInfoEx(csbiex)` in `adaptDispatch`. Wonder how to do the same in `_terminalApi.`

**Comment by zadjii-msft** [comment]
@jsoref What the heck happened with the spellcheck here? `noexcept` is _definitely_ a word... Weird.

File: src/cascadia/TerminalCore/TerminalDispatch.cpp
Comment (zadjii-msft) [reply src/cascadia/TerminalCore/TerminalDispatch.cpp:148]: meh, I'd just add a `til::point GetTerminalSize()` to `ITerminalApi` (or just make it return a `COORD` if the `til` type doesn't work in that header)

File: src/cascadia/TerminalCore/TerminalDispatch.cpp
Comment (zadjii-msft) [inline src/cascadia/TerminalCore/TerminalDispatch.cpp:139]: technically, this should be a `false` since it's unimplemented, yea?

File: src/cascadia/TerminalCore/TerminalDispatch.cpp
Comment (zadjii-msft) [inline src/cascadia/TerminalCore/TerminalDispatch.cpp:177]: This definitely is out of scope for this PR, but I wonder if there isn't some way that we could have a common implementation for both `AdaptDispatch` and `TerminalDispatch` for these methods. Since the `_tabStops` are stored internal to the dispatcher, it seems like both of them could somehow have a common implementation...

**Review by zadjii-msft — CHANGES_REQUESTED** [review]
I'm mostly blocking because we should probably have some unittests for this. I'm sure we've already got something similar in the `ScreenBufferTests`, just a matter of copying them to the terminal core unit tests

**Comment by jsoref** [comment]
I'll have to look tonight, it's currently in expect, it should probably move to dictionary/api since you're right, it's a word!

But that doesn't explain why the bot doesn't like it... https://github.com/microsoft/terminal/blob/691e02ef1ca2952304b0e79222fbfcad7bd98b72/.github/actions/spelling/expect/expect.txt#L1535

**Comment by jsoref** [comment]
@zadjii-msft: So, the log output is interesting:
> sed: can't read /tmp/spelling/expect.words.txt: No such file or directory

Effectively, it ended up w/o _any_ expect list. 

And the reason you didn't see more is:
>       "message": "Body is too long (maximum is 65536 characters)"

It was going to complain about *everything*, but you were saved by the GitHub character limit. Now, the question is: "why?" Is it possible that this branch needs a rebase?

And the answer is: yes.

This is from when I renamed the directory from `.github/actions/spell-check` to `.github/actions/spelling`.

Either rebase or force a merge.

Sorry about this.

**Comment by zadjii-msft** [comment]
@jsoref no worries! I thought it might be something like that. A merge/rebase is easy enough ☺️ (over to you @skyline75489)

File: src/cascadia/TerminalCore/TerminalDispatch.cpp
Comment (j4james) [reply src/cascadia/TerminalCore/TerminalDispatch.cpp:177]: FYI, my long term plan is to replace the `TerminalDispatch` class entirely with `AdaptDispatch`, so the common implementation in this case will ultimately just be `AdaptDispatch`. For now, though, I've been trying to keep the `TerminalDispatch` implementation as close as possible to the corresponding `AdaptDispatch` code, just with `_pConApi` changed to `_terminalApi`. I'm hoping that'll make things easier to synchronize when we eventually merge the two.

Comment (skyline75489) [reply src/cascadia/TerminalCore/TerminalDispatch.cpp:177]: @j4james your plan makes sense. Would you still like me to continue the work here? Or can you point me the correct direction towards merging these two?

Comment (j4james) [reply src/cascadia/TerminalCore/TerminalDispatch.cpp:177]: There's a ton of work that needs to be done on the conapi interface before we can merge. Other than that, we need to make sure that everything that's currently only implemented in `TerminalDispatch` is also copied across to `AdaptDispatch`. And as I said above, if you're copying functionality from `AdaptDispatch` (like these tabs), it's preferable if the code is essentially identical between the two.

So as long as you're doing that, I don't think there's anything wrong with this PR. I'm just not sure why we actually need the functionality in `TerminalDispatch` at this point. Are we intending to pass these tab sequences through conpty?

Comment (zadjii-msft) [reply src/cascadia/TerminalCore/TerminalDispatch.cpp:177]: Oh, I like that plan. That would be nice. I knew this would be out-of-scope for this PR, but this is a good vision to have.

I think the bigger vision here is @skyline75489 is working on tmux control mode, and that needs much more complete VT support than `TerminalDispatch` offers right now. Eventually we'll want it for passthrough mode, but that's not really in the cards right now.

**Comment by skyline75489** [comment]
Sorry for the confusion. I talked to the core devs internally about this but forgot to mention it in this PR. I’m trying to make tmux CC happen by directly passing the output to terminal.

获取 Outlook for iOS<https://aka.ms/o0ukef>
________________________________
发件人: James Holderness ***@***.***>
发送时间: Tuesday, March 16, 2021 6:39:25 PM
收件人: microsoft/terminal ***@***.***>
抄送: Chester Liu ***@***.***>; Mention ***@***.***>
主题: Re: [microsoft/terminal] [WIP] Initial implementation for tab stops in TerminalDispatch (#9444)


@j4james commented on this pull request.

________________________________

In src/cascadia/TerminalCore/TerminalDispatch.cpp<https://github.com/microsoft/terminal/pull/9444#discussion_r595047812>:

> +    while (column > 0 && tabsPerformed < numTabs)
+    {
+        column--;
+        if (til::at(_tabStopColumns, column))
+        {
+            tabsPerformed++;
+        }
+    }


There's a ton of work that needs to be done on the conapi interface before we can merge. Other than that, we need to make sure that everything that's currently only implemented in TerminalDispatch is also copied across to AdaptDispatch. And as I said above, if you're copying functionality from AdaptDispatch (like these tabs), it's preferable if the code is essentially identical between the two.

So as long as you're doing that, I don't think there's anything wrong with this PR. I'm just not sure why we actually need the functionality in TerminalDispatch at this point. Are we intending to pass these tab sequences through conpty?

―

**Comment by skyline75489** [comment]
Knowing @j4james 's vision of merging `AdaptDispatch` & `TerminalDispath`, I realize I'm doing this the inverse way. I'll close this and see if I can tweak things inside `AdaptDispatch` instead.

**Pull request closed without merging**
